src/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.PowMod.cs

```diff
@@ -0,0 +1,432 @@
+﻿using System.Diagnostics;
+using System.Security;
+
+namespace System.Numerics
+{
+    internal static partial class BigIntegerCalculator
+    {
+        // Executes different exponentiation algorithms, which are
+        // basically based on the classic square-and-multiply method.
+
+        // https://en.wikipedia.org/wiki/Exponentiation_by_squaring
+
+        public static uint[] Pow(uint value, uint power)
+        {
+            // The basic pow method for a 32-bit integer.
+            // To spare memory allocations we first roughly
+            // estimate an upper bound for our buffers.
+
+            int length = PowBound(power, 1, 1);
+
+            BitsBuffer tempBuffer = new BitsBuffer(length, 0);
+            BitsBuffer valueBuffer = new BitsBuffer(length, value);
+            BitsBuffer resultBuffer = new BitsBuffer(length, 1);
+
+            PowCore(power, ref valueBuffer, ref resultBuffer,
+                ref tempBuffer);
+
+            return resultBuffer.ToArray();
+        }
+
+        public static uint[] Pow(uint[] value, uint power)
+        {
+            Debug.Assert(value != null);
+
+            // The basic pow method for a big integer.
+            // To spare memory allocations we first roughly
+            // estimate an upper bound for our buffers.
+
+            int length = PowBound(power, value.Length, 1);
+
+            BitsBuffer tempBuffer = new BitsBuffer(length, 0);
+            BitsBuffer valueBuffer = new BitsBuffer(length, value);
+            BitsBuffer resultBuffer = new BitsBuffer(length, 1);
+
+            PowCore(power, ref valueBuffer, ref resultBuffer,
+                ref tempBuffer);
+
+            return resultBuffer.ToArray();
+        }
```

Comment:
The two Pow methods are almost identical.  I wonder if they should be refactored to share most of the code, e.g.
```C#
public static uint[] Pow(uint value, uint power)
{
    int length = PowBound(power, 1, 1);
    BitsBuffer valueBuffer = new BitsBuffer(length, value);
    return PowCore(length, power, ref valueBuffer);
}

public static uint[] Pow(uint[] value, uint power)
{
    Debug.Assert(value != null);
    int length = PowBound(power, value.Length, 1);
    BitsBuffer valueBuffer = new BitsBuffer(length, value);
    return PowCore(length, power, ref valueBuffer);
}

private static uint[] PowCore(int length, uint power, ref BitsBuffer valueBuffer)
{
    BitsBuffer tempBuffer = new BitsBuffer(length, 0);
    BitsBuffer resultBuffer = new BitsBuffer(length, 1);
    PowCore(power, ref valueBuffer, ref resultBuffer, ref tempBuffer);
    return resultBuffer.ToArray();
}
```
or something like that?